opentracing/src/main/java/opentracing/Tracer.java

```diff
@@ -0,0 +1,94 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+
+/**
+ * Tracer is a simple, thin interface for Span creation, and Span propagation into different transport formats.
+ */
+public interface Tracer {
+
+  /**
+   * Create, start, and return a new Span with the given `operationName`.
+   * An optional parent Span can be specified used to incorporate the newly-returned Span into an existing trace.
+   *
+   * <p>Example:
+   * <pre>{@code
+    Tracer tracer = ...
+
+    Span feed = tracer.buildSpan("GetFeed")
+                      .start();
+
+    Span http = tracer.buildSpan("HandleHTTPRequest")
+                      .withParent(feed)
+                      .withTag("user_agent", req.UserAgent)
+                      .withTag("lucky_number", 42)
+                      .start();
+    }</pre>
+   */
+  SpanBuilder buildSpan(String operationName);
+
+  /** Takes two arguments:
+   *    a Span instance, and
+   *    a “carrier” object in which to inject that Span for cross-process propagation.
+   *
+   */
+  <T> void inject(Span span, T carrier);
+
+  /**  Returns a SpanBuilder provided
+   *    a “carrier” object from which to extract identifying information needed by the new Span instance.
+   *
+   * If the carrier object has no such span stored within it, a new Span is created.
+   *
+   * Unless there’s an error, it returns a SpanBuilder.
+   * The Span generated from the builder can be used in the host process like any other.
+   *
+   * (Note that some OpenTracing implementations consider the Spans on either side of an RPC to have the same identity,
+   * and others consider the caller to be the parent and the receiver to be the child)
+   */
+  <T> SpanBuilder join(T carrier);
+
+  interface SpanBuilder {
+
+      /** Specify the operationName.
+       *
+       * If the operationName has already been set (implicitly or explicitly) an IllegalStateException will be thrown.
+       */
+      SpanBuilder withOperationName(String operationName);
+
+      /** Specify the parent span
+       *
+       * If the parent has already been set an IllegalStateException will be thrown.
+       */
+      SpanBuilder withParent(Span parent);
+
+      /** Specify a timestamp the Span actually started from.
+       *
+       * If the timestamp has already been set an IllegalStateException will be thrown.
+       */
+      SpanBuilder withTimestamp(long microseconds);
+
+      /** Same as {@link Span#setTag(String, String)}, but for the span being built. */
+      SpanBuilder withTag(String key, String value);
+
+      /** Same as {@link Span#setTag(String, String)}, but for the span being built. */
+      SpanBuilder withTag(String key, boolean value);
+
+      /** Same as {@link Span#setTag(String, String)}, but for the span being built. */
+      SpanBuilder withTag(String key, Number value);
+
+      /** Returns the started Span. */
+      Span start();
```

Comment:
> so would we add start(parent) too?

No. I believe @adriancole suggested this, and why it made sense to me, is because start and timestamp are close concepts. having a choice between calling `start()` or `start(timestamp)` intuitively makes sense. ie start now or start at this timestamp. that's part of designing a fluent api.

> Starting a child span is by far the most frequent usage …

It's not what's frequent, but what intuitively makes sense or reads well, ie fluent api.

That said, it could be argued that `start(timestamp)` isn't intuitive enough, and i've no strong opinion to that. `withTimestamp(..)` is slightly more intuitive if you ask me. so i'd be happy to return the code if this was the argument raised.